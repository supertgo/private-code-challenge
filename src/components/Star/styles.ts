import styled, { css, DefaultTheme } from 'styled-components';
import media from 'styled-media-query';

import { StartProps } from '.';

type WrapperProps = Pick<StartProps, 'color' | 'size'>;

const wrapperModiffiers = {
  small: () => css`
    width: 8rem;
    height: 8rem;
  `,
  medium: () => css`
    width: 10rem;
    height: 10rem;
  `,
  primary: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
  `,
  secundary: (theme: DefaultTheme) => css`
    color: ${theme.colors.primary};
  `
};

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, color, size }) => css`
    display: flex;
    background-color: ${theme.colors[color!]};
    position: relative;
    text-align: center;
    justify-content: center;

    ${size && wrapperModiffiers[size]()}
    ${!!color && wrapperModiffiers[color](theme)}

    &:before {
      content: '';
      position: absolute;
      background-color: ${theme.colors[color!]};
      transform: rotate(30deg);
      ${size && wrapperModiffiers[size]()};
    }

    &:after {
      content: '';
      position: absolute;
      background-color: ${theme.colors[color!]};
      transform: rotate(-30deg);
      ${size && wrapperModiffiers[size]()};
    }
  `}
`;

export const Content = styled.div`
  ${({ theme }) => css`
    align-self: center;
    transform: rotate(20deg);
    z-index: ${theme.layers.base};
  `}
`;

export const Promotional = styled.span`
  ${({ theme }) => css`
    align-self: center;
    font-size: ${theme.font.sizes.xlarge};
    font-weight: 800;
    z-index: ${theme.layers.base};

    ${media.greaterThan('medium')`
      font-size: 4rem;
    `}
  `}
`;

export const Text = styled.p`
  ${({ theme }) => css`
    font-weight: ${theme.font.light};
    font-size: ${theme.font.sizes.xxsmall};
    z-index: ${theme.layers.base};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.medium};
    `}
  `}
`;
