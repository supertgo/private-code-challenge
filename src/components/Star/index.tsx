import * as S from './styles';

export type StartProps = {
  color?: 'primary' | 'secundary';
  size?: 'small' | 'medium';
  children?: string;
};

const Star = ({
  color = 'secundary',
  size = 'medium',
  children
}: StartProps) => (
  <S.Wrapper aria-label="star" color={color} size={size}>
    <S.Content>
      <S.Promotional>{children}</S.Promotional>
      <S.Text>de desconto</S.Text>
    </S.Content>
  </S.Wrapper>
);

export default Star;
