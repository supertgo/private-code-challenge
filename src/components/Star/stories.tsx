import { Story, Meta } from '@storybook/react';
import Star, { StartProps } from '.';

export default {
  title: 'Star',
  component: Star
} as Meta;

export const Default: Story<StartProps> = (args) => (
  <div
    style={{
      maxWidth: '104rem',
      margin: '0 auto',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    }}
  >
    <Star {...args} />
  </div>
);

Default.args = {
  children: '70%'
};
