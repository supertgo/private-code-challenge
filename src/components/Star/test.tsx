import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helper';

import Star from '.';

describe('<Star />', () => {
  it('should render the default configs', () => {
    const { container } = renderWithTheme(<Star>70%</Star>);

    expect(screen.getByText('70%')).toBeInTheDocument();
    expect(screen.getByLabelText(/star/i)).toHaveStyle({
      backgroundColor: '#3AFFE5',
      width: '10rem',
      height: '10rem'
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the primary color when passed', () => {
    renderWithTheme(<Star color="primary">70%</Star>);
    expect(screen.getByLabelText(/star/i)).toHaveStyle({
      backgroundColor: '#252160'
    });
  });

  it('it should render the small start when passed small', () => {
    renderWithTheme(<Star size="small">70%</Star>);
    expect(screen.getByLabelText(/star/i)).toHaveStyle({
      width: '8rem',
      height: '8rem'
    });
  });
});
