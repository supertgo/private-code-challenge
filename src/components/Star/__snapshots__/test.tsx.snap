// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Star /> should render the default configs 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #3AFFE5;
  position: relative;
  text-align: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 10rem;
  height: 10rem;
  color: #252160;
}

.c0:before {
  content: '';
  position: absolute;
  background-color: #3AFFE5;
  -webkit-transform: rotate(30deg);
  -ms-transform: rotate(30deg);
  transform: rotate(30deg);
  width: 10rem;
  height: 10rem;
}

.c0:after {
  content: '';
  position: absolute;
  background-color: #3AFFE5;
  -webkit-transform: rotate(-30deg);
  -ms-transform: rotate(-30deg);
  transform: rotate(-30deg);
  width: 10rem;
  height: 10rem;
}

.c1 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transform: rotate(20deg);
  -ms-transform: rotate(20deg);
  transform: rotate(20deg);
  z-index: 10;
}

.c2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  font-size: 2.0rem;
  font-weight: 800;
  z-index: 10;
}

.c3 {
  font-weight: 300;
  font-size: 0.8rem;
  z-index: 10;
}

@media (min-width:768px) {
  .c2 {
    font-size: 4rem;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 1.6rem;
  }
}

<div
  aria-label="star"
  class="c0"
  color="secundary"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    >
      70%
    </span>
    <p
      class="c3"
    >
      de desconto
    </p>
  </div>
</div>
`;
